samza-core/src/main/scala/org/apache/samza/container/TaskInstance.scala

```diff
@@ -77,97 +78,112 @@ class TaskInstance(
 
   def registerMetrics {
-    debug("Registering metrics for taskName: %s" format taskName)
+    debug("Registering metrics for task name: %s" format taskName)
 
     reporters.values.foreach(_.register(metrics.source, metrics.registry))
   }
 
   def registerOffsets {
-    debug("Registering offsets for taskName: %s" format taskName)
+    debug("Registering offsets for task name: %s" format taskName)
 
-    offsetManager.register(taskName, systemStreamPartitions)
+    val sspsToRegister = systemStreamPartitions -- sideInputSSPs
+    offsetManager.register(taskName, sspsToRegister)
   }
 
   def startStores {
     if (storageManager != null) {
-      debug("Starting storage manager for taskName: %s" format taskName)
+      debug("Starting storage manager for task name: %s" format taskName)
 
       storageManager.init
     } else {
-      debug("Skipping storage manager initialization for taskName: %s" format taskName)
+      debug("Skipping storage manager initialization for task name: %s" format taskName)
     }
   }
 
   def startTableManager {
     if (tableManager != null) {
-      debug("Starting table manager for taskName: %s" format taskName)
+      debug("Starting table manager for task name: %s" format taskName)
 
       tableManager.init(containerContext, context)
     } else {
-      debug("Skipping table manager initialization for taskName: %s" format taskName)
+      debug("Skipping table manager initialization for task name %s" format taskName)
     }
   }
 
   def initTask {
     if (isInitableTask) {
-      debug("Initializing task for taskName: %s" format taskName)
+      debug("Initializing task for task name %s" format taskName)
 
       task.asInstanceOf[InitableTask].init(config, context)
     } else {
-      debug("Skipping task initialization for taskName: %s" format taskName)
+      debug("Skipping task initialization for task name %s" format taskName)
     }
   }
 
   def registerProducers {
-    debug("Registering producers for taskName: %s" format taskName)
+    debug("Registering producers for task name %s" format taskName)
 
     collector.register
   }
 
   def registerConsumers {
-    debug("Registering consumers for taskName: %s" format taskName)
+    debug("Registering consumers for task name %s" format taskName)
 
     systemStreamPartitions.foreach(systemStreamPartition => {
-      val offset = offsetManager.getStartingOffset(taskName, systemStreamPartition)
-      .getOrElse(throw new SamzaException("No offset defined for SystemStreamPartition: %s" format systemStreamPartition))
-      consumerMultiplexer.register(systemStreamPartition, offset)
-      metrics.addOffsetGauge(systemStreamPartition, () => {
-        offsetManager
-          .getLastProcessedOffset(taskName, systemStreamPartition)
-          .orNull
-      })
+      var metricsValueGetter: () => String = null
+      var offset: Option[String] = None
+
+      if (sideInputSSPs.contains(systemStreamPartition)) {
```

Comment:
It seems like it might be difficult to ensure that this pattern of:
"if (is side input) {
// use storageManager.getStartingOffset
} else {
// use offsetManager.getStartingOffset
}" is used correctly. It seems like offsetManager won't work on side input SSPs, but there is nothing stopping someone from using a side input SSP with offsetManager except if they know they need to check it is a side input SSP. The system stream partitions are even available through TaskContext, so application code has access to them (maybe even for custom checkpointing?). In addition, this if-else pattern (also used further below) seems a bit hard to maintain.

Is there a way to more strictly enforce that side input SSPs have to use storageManager.getStartingOffset?
Maybe if there was a way to more cleanly categorize an SSP as a side input or not, then you wouldn't need these checks.